modules/platforms/cpp/ignite/client/detail/cancellation_token_impl.cpp

```diff
@@ -0,0 +1,121 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements. See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License. You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "ignite/client/detail/cancellation_token_impl.h"
+
+#include <sstream>
+
+namespace ignite
+{
+
+void cancellation_token_impl::cancel_async(ignite_callback<void> callback) {
+    std::lock_guard guard(m_mutex);
+
+    if (is_cancelled()) {
+        if (m_result) {
+            callback(ignite_result<void>{*m_result});
+        } else {
+            m_callbacks.push_back(std::move(callback));
+        }
+        return;
+    }
+
+    m_cancelled = true;
+    m_callbacks.push_back(std::move(callback));
+
+    if (m_actions.empty()) {
+        m_result = ignite_result<void>{};
+        for (auto &cb : m_callbacks) {
+            cb(ignite_result<void>{*m_result});
+        }
+        return;
+    }
+
+    auto results = std::make_shared<std::vector<ignite_result<void>>>();
+    auto results_mutex = std::make_shared<std::mutex>();
+    auto expected_results = m_actions.size();
+    for (auto &action : m_actions) {
+        action([this, results, results_mutex, expected_results] (ignite_result<void> res) {
+            bool all_done{false};
```

Comment:
We are capturing `this` here and we use some mutex to protect the results list, so I take it that this action callback might be called concurrently/asynchronously. How do we make sure that `this` lives long enough? Since we inherit from `enabled_shared_from_this` I think we should capture a `shared_ptr` instead.

Reply:
Good catch, it should have been `shared_from_this()` all along.